pkg/relayer/session/persistence.go

```diff
@@ -0,0 +1,200 @@
+			continue
+		}
+
+		sessionEndHeight := persistedSMT.SessionHeader.SessionEndBlockHeight
+		sessionId := persistedSMT.SessionHeader.SessionId
+		supplierOperatorAddress := persistedSMT.SupplierOperatorAddress
+
+		// Create a contextual logger for this specific session
+		sessionLogger := logger.With(
+			"session_id", sessionId,
+			"supplier_operator_address", supplierOperatorAddress,
+			"session_end_height", sessionEndHeight,
+		)
+
+		// Session lifecycle states and corresponding actions:
+		// 1. Active and accepting relays: Retain session for processing new relays
+		// 2. Within claim window: Retain session as it's in the process of creating a claim
+		// 3. Past claim window with onchain claim: Retain for proof submission
+		// 4. Past claim window without onchain claim: Delete as it's too late for settlement
+		// 5. Past proof window: Delete as the session is completely expired
+
+		claimWindowCloseHeight := sharedtypes.GetClaimWindowCloseHeight(sharedParams, sessionEndHeight)
+		proofWindowCloseHeight := sharedtypes.GetProofWindowCloseHeight(sharedParams, sessionEndHeight)
+
+		// Check if session is expired (past proof submission window)
+		// These sessions can no longer reach settlement and should be deleted
+		if height > proofWindowCloseHeight {
+			if storeErr := rs.deletePersistedSessionTree(persistedSMT); storeErr != nil {
+				sessionLogger.Error().Err(storeErr).Msg("failed to delete outdated session tree, skipping")
+				continue
+			}
+
+			sessionLogger.Info().
+				Int64("proof_window_close", proofWindowCloseHeight).
+				Int64("current_height", height).
+				Msg("Session deleted: expired past proof submission window")
+
+			continue
+		}
+
+		// For sessions past the claim window, check if a claim exists onchain
+		// If no claim exists, the session can't progress to settlement and should be deleted
+		var claim *prooftypes.Claim
+		if height > claimWindowCloseHeight {
+			claim, queryErr := rs.proofQueryClient.GetClaim(ctx, supplierOperatorAddress, sessionId)
+			// No claim was created onchain, delete the session from the store.
+			if queryErr != nil || claim == nil {
+				if storeErr := rs.deletePersistedSessionTree(persistedSMT); storeErr != nil {
+					sessionLogger.Error().Err(storeErr).Msg("failed to delete outdated session without claim, skipping")
+					continue
+				}
+
+				sessionLogger.Info().Msg("deleting session has no claim created onchain, deleting session metadata")
+				continue
+			}
+		}
+
+		// The session has still a chance to reach settlement.
+		// Add it to the sessionTrees map of blockHeight->sessionId->supplierOperatorAddress->sessionTree
+		sessionTreesWithEndHeight, ok := rs.sessionsTrees[sessionEndHeight]
+
+		// If there is no map for sessions at the sessionEndHeight, create one.
+		if !ok {
+			sessionTreesWithEndHeight = make(map[string]map[string]relayer.SessionTree)
+			rs.sessionsTrees[sessionEndHeight] = sessionTreesWithEndHeight
+		}
```

Comment:
I feel like this should be extracted to a method. Do you think there are other places where this may be duplicated?